### PR TITLE
fix(route): 通过在配置中预添加大量 Cookie 来绕过 Bilibili 反爬

### LINKED ISSUE #13315: bilibili 401 Error message: Got error code -401
### 路由地址

```routes
/bilibili/user/video/63231
```

### 完整路由地址

```fullroutes
无
```

### 相关文档

https://docs.rsshub.app/zh/routes/social-media#bilibili-up-zhu-tou-gao

### 预期是什么？

正常获取

### 实际发生了什么？

401错误

### 部署

自建

### 部署相关信息

OS:Linux Docker: 24.0.5 

### 额外信息

```shell
无
```

### 这不是重复的 issue

- [X] 我已经搜索了 [现有 issue](https://github.com/DIYgod/RSSHub/issues)，以确保该错误尚未被报告。

### LINKED ISSUE #13334: bilibili用户关注动态title显示undefined
### 路由地址

```routes
/bilibili/followings/dynamic/:uid/:routeParams?
```

### 完整路由地址

```fullroutes
/bilibili/followings/dynamic/:uid/:routeParams?
```

### 相关文档

https://docs.rsshub.app/zh/routes/social-media#bilibili-yong-hu-guan-zhu-dong-tai

### 预期是什么？

显示正确的title

### 实际发生了什么？

显示undefined
image

### 部署

自建

### 部署相关信息

_No response_

### 额外信息

```shell
无
```

### 这不是重复的 issue

- [X] 我已经搜索了 [现有 issue](https://github.com/DIYgod/RSSHub/issues)，以确保该错误尚未被报告。

### PR DESCRIPTION
**Pull request by NightSpaceC**
## Involved Issue / 该 PR 相关 Issue

Close #13334
Close #13315

## Example for the Proposed Route(s) / 路由地址示例

```routes
/bilibili/user/video/25151282
/bilibili/user/video-all/25151282
```

## New RSS Route Checklist / 新 RSS 路由检查表
  
- [ ] New Route / 新的路由
  - [ ] Follows [v2 Script Standard](https://docs.rsshub.app/joinus/advanced/script-standard) / 跟随 [v2 路由规范](https://docs.rsshub.app/zh/joinus/advanced/script-standard)
- [ ] Documentation / 文档说明
  - [ ] EN / 英文文档
  - [ ] CN / 中文文档
- [ ] Full text / 全文获取
  - [ ] Use cache / 使用缓存
- [ ] Anti-bot or rate limit / 反爬/频率限制 
  - [ ] If yes, do your code reflect this sign? / 如果有, 是否有对应的措施? 
- [ ] [Date and time](https://docs.rsshub.app/joinus/advanced/pub-date) / [日期和时间](https://docs.rsshub.app/zh/joinus/advanced/pub-date)
  - [ ] Parsed / 可以解析
  - [ ] Correct time zone / 时区正确
- [ ] New package added / 添加了新的包
- [ ] `Puppeteer`

## Note / 说明

在访问频率超出限制时，Bilibili 对单个会话要求输入验证码而非同一 IP 的所有会话

现在获取数据时随机返回配置中的一个 Cookie 用于查询来降低单 Cookie 访问频率

预添加的 Cookie 可以是游客状态

可以使用类似于

```py
import sys
import time
import requests_html


def cookieToString(cookie):
    l = []
    for each in cookie.get_dict().items():
        l.append(each[0] + '=' + each[1] + '; ')
    return ''.join(l)

number = int(sys.argv[1])
for i in range(0, number):
    session = requests_html.HTMLSession()
    res = session.get('https://www.bilibili.com', headers={'user-agent': 'Your UA'})
    print('set BILIBILI_COOKIE_data%d=%s' % (i, cookieToString(res.cookies)))
    time.sleep(0.2)
```

的脚本来大量获取 Cookie

值得注意的是，当请求频率非常大时，此方法会失效，因为此时这个 IP 下的所有会话都会被要求输入验证码

### CODE REVIEW

File: lib/v2/bilibili/cache.js [hunk 1]
```diff
@@ -1,19 +1,19 @@
 const got = require('@/utils/got');
 const utils = require('./utils');
 const cheerio = require('cheerio');
+const config = require('@/config').value;
 
 module.exports = {
     getCookie: (ctx) => {
+        if (Object.keys(config.bilibili.cookies).length > 0) {
```
Comment (TonyRL) [inline lib/v2/bilibili/cache.js:8]: I guess `getCookie` needs an update. Maybe filling `buvid3`, `b_nut`, `b_lsid`, `_uuid`, `buvid4` cookies when performing API requests will have less strict rate limit.

1. `buvid3` and `b_nut` will be provided by server when you visited the site first time, either `https://www.bilibili.com/` or `https://space.bilibili.com/1` will have them.

<details><summary>2. compute <code>b_lsid</code> and <code>_uuid</code></summary>

```js
// a
function randomHexStr(length) {
    let string = '';
    for (let r = 0; r < length; r++) {
        string += dec2HexUpper(16 * Math.random());
    }
    return padStringWithZeros(string, length);
}
// o
function dec2HexUpper(e) {
    return Math.ceil(e).toString(16).toUpperCase();
}
// s
function padStringWithZeros(string, length) {
    let padding = '';
    if (string.length < length) {
        for (let n = 0; n < length - string.length; n++) {
            padding += '0';
        }
    }
    return padding + string;
}
function lsid() {
    const e = Date.now().toString(16).toUpperCase();
    const lsid = randomHexStr(8) + '_' + e;
    return lsid;
}
function _uuid() {
    let e = randomHexStr(8);
    let t = randomHexStr(4);
    let r = randomHexStr(4);
    let n = randomHexStr(4);
    let o = randomHexStr(12);
    let i = Date.now();
    return e + '-' + t + '-' + r + '-' + n + '-' + o + padStringWithZeros((i % 100000).toString(), 5) + 'infoc';
}
```

</details>

3. GET `api.bilibili.com/x/frontend/finger/spi` with cookies `buvid3`, `b_nut`, `b_lsid`, `_uuid` which will return `buvid4` from `data.b_4`

4. continuing typical api request like `/x/space/wbi/acc/info` or `/x/space/wbi/arc/search`

Note: The value of the cookie `buvid_fp` and `fingerprint` are basically the output of [fingerprintjs](https://github.com/fingerprintjs/fingerprintjs)

**Comment by TonyRL** [comment]
For future reference, this is how `bili_ticket` and `bili_ticket_expires` are generated.

```js
const e = Math.floor(Date.now() / 1000);
// P
function shiftCharByOne(string) {
    let shiftedStr = '';
    for (let n = 0; n < string.length; n++) {
        shiftedStr += String.fromCharCode(string.charCodeAt(n) - 1);
    }
    return shiftedStr;
}
// o
function hexsign() {
    const n = 'YhxToH[2q';
    const r = CryptoJS.HmacSHA256('ts'.concat(e), shiftCharByOne(n));
    const o = CryptoJS.enc.Hex.stringify(r);
    return o;
}
```

POST `https://api.bilibili.com/bapis/bilibili.api.ticket.v1.Ticket/GenWebTicket`
Query String Parameter:
`key_id`: ec02
`hexsign`: hexsign()
`context[ts]`: e 
`csrf`: 

The value of `bili_ticket` will be `data.ticket` (a JWT) and `bili_ticket_expires` will be `data.created_at` + `data.ttl`.

---

Since `data.nav.img` and `data.nav.sub` also here, there might be a possibility that `/x/web-interface/nav` will be removed, but who knows.

**Comment by NightSpaceC** [comment]
I've done what you said, but the part about fingerprinting I can't do because I can't use that library in Node

Unfortunately, doing so doesn't seem to result in looser rate limits, at least on my computer

**Comment by NightSpaceC** [comment]
After testing, video-all may still be detected, but this cannot be solved because the verification code also appeared when I quickly jumped to the page many times manually by using the browser.

**Comment by TonyRL** [comment]
I don't expect /video-all to work at all since it's way too aggressive.

**Pull request merged**